### DISCUSSION TITLE
Need help with resolving an "ETIMEDOUT" error on ESP32 with Micropython

### DISCUSSION THREAD

**Discussion by OMOLOYIN1** [micropython/micropython#10114] [Using MicroPython]
I am using an ESP32 board and a MPU6050 sensor. My software version is Micropython v1.18. I connected the circuit by powering the sensor with 3.3 volts from the ESP32. I also used  I2c protocol by connecting the SCL and SDA pins of the ESP32 to the SCL and SDA pins of the MPU6050 respectively. My setup worked well for days. I was able to read the accelerometer and gyroscope values. Suddenly, my code stopped working. And I got the following error. And I have battling with this problem for over a week. I will appreciate any help

```
Traceback (most recent call last):
  File "main.py", line 75, in <module>
  File "main.py", line 36, in mpu6050_init
OSError: [Errno 116] ETIMEDOUT
```

My code is as follows:

```
#import PIN and I2C from machine library 
from machine import Pin, SoftI2C,
import time

#Define I2C bus 
i2c = SoftI2C(sda=Pin(21), scl=Pin(22))

#Device address on the I2C bus
MPU6050_ADDR = 0x68

#PWR_MGMT_1 memory address
MPU6050_PWR_MGMT_1 = 0x6B

#Accelerometer and Gyroscope's high and low register for each axis
MPU6050_ACCEL_XOUT_H = 0x3B
MPU6050_ACCEL_XOUT_L = 0x3C
MPU6050_ACCEL_YOUT_H = 0x3D
MPU6050_ACCEL_YOUT_L = 0x3E
MPU6050_ACCEL_ZOUT_H = 0x3F
MPU6050_ACCEL_ZOUT_L = 0x40
MPU6050_GYRO_XOUT_H = 0x43
MPU6050_GYRO_XOUT_L = 0x44
MPU6050_GYRO_YOUT_H = 0x45
MPU6050_GYRO_YOUT_L = 0x46
MPU6050_GYRO_ZOUT_H = 0x47
MPU6050_GYRO_ZOUT_L = 0x48

#Accelerometer's LSB/g (least significant bits per gravitational force) sensitivity
MPU6050_LSBG = 16384.0

#Gyroscope's LSB/g sensitivity
MPU6050_LSBDS = 131.0 


#Set all bits in the PWR_MGMT_1 register to 0
def mpu6050_init(i2c):
    i2c.writeto_mem(MPU6050_ADDR, MPU6050_PWR_MGMT_1, bytes([0])) #LINE 36


def combine_register_values(h, l):
    if not h[0] & 0x80:
        return h[0] << 8 | l[0]
    return -((h[0] ^ 255) << 8) |  (l[0] ^ 255) + l


#Get Accelerometer values
def mpu6050_get_accel(i2c):
    accel_x_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_XOUT_H, 1)
    accel_x_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_XOUT_L, 1)
    accel_y_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_YOUT_H, 1)
    accel_y_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_YOUT_L, 1)
    accel_z_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_ZOUT_H, 1)
    accel_z_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_ACCEL_ZOUT_L, 1)
    
    return [combine_register_values(accel_x_h, accel_x_l) / MPU6050_LSBG,
            combine_register_values(accel_y_h, accel_y_l) / MPU6050_LSBG,
            combine_register_values(accel_z_h, accel_z_l) / MPU6050_LSBG]


#Get Gyroscope values
def mpu6050_get_gyro(i2c):
    gyro_x_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_XOUT_H, 1)
    gyro_x_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_XOUT_L, 1)
    gyro_y_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_YOUT_H, 1)
    gyro_y_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_YOUT_L, 1)
    gyro_z_h = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_ZOUT_H, 1)
    gyro_z_l = i2c.readfrom_mem(MPU6050_ADDR, MPU6050_GYRO_ZOUT_L, 1)
    
    return [combine_register_values(gyro_x_h, gyro_x_l) / MPU6050_LSBDS,
            combine_register_values(gyro_y_h, gyro_y_l) / MPU6050_LSBDS,
            combine_register_values(gyro_z_h, gyro_z_l) / MPU6050_LSBDS]


if __name__ == "__main__":
    i2c = SoftI2C(sda=Pin(21), scl=Pin(22))
    mpu6050_init(i2c) #LINE 75
    
    while True:
        print("Accelerometer:\t", mpu6050_get_accel(i2c), "g") #Print Accelerometer values (X,Y,Z) 
        #print("Gyroscope:\t", mpu6050_get_gyro(i2c), "°/s") #Print Gyroscope values (X,Y,Z)
        time.sleep(1)
```

**Comment by karfas** [comment]
ETIMEDOUT most likely means that the device isn't found on the i2c bus.

You might want to check
- the original line in the driver file
- cables
- pull up resistors on the bus
- does i2c.scan() find your device ?
- does i2c.scan() find other devices (you might want to test with other i2c _things_.

**Reply by OMOLOYIN1** [reply]
I did an I2C scan a few hours ago. No device was found. Therefore, I replaced with the MPU6050 with a new one. And the setup started working again. Then, I re-tried the old MPU6050. Then, the old sensor also worked. But, it crashed after producing only 4 readings.

I'm thinking that I have damaged the old/previous sensor one way or the other. But I can say this with absolute assurance.

**Reply by peterhinch** [reply]
This does hint at an electrical problem as @karfas suggested. In addition to his suggestions, ensure wires are short and there is a short, direct, gnd wire between ESP32 and MPU6050. Another thing that can go wrong with WiFi connected chips is that when they transmit to WiFi the current consumption goes up. This can cause voltage drops on the USB lead (if that's how you're powering it). Use a short, good quality, USB lead.

**Comment by peterhinch** [comment]
You might also like to try [this driver](https://github.com/micropython-IMU/micropython-mpu9x50): the base class in `imu.py` supports MPU6050 and the driver uses techniques that reduce RAM allocations.

**Reply by OMOLOYIN1** [reply]
Thank you sir. I'll look into that too.
